scripts/build_osm_network.py

```diff
@@ -0,0 +1,1051 @@
+# -*- coding: utf-8 -*-
+# SPDX-FileCopyrightText: : 2020-2024 The PyPSA-Eur and PyPSA-Earth Authors
+#
+# SPDX-License-Identifier: MIT
+
+import logging
+import os
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+from _benchmark import memory_logger
+from _helpers import configure_logging, set_scenario_config
+from shapely.geometry import LineString, Point
+from shapely.ops import linemerge, split
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+# list of recognised nan values (NA and na excluded as may be confused with Namibia 2-letter country code)
+NA_VALUES = ["NULL", "", "N/A", "NAN", "NaN", "nan", "Nan", "n/a", "null"]
+
+
+def read_csv_nafix(file, **kwargs):
+    "Function to open a csv as pandas file and standardize the na value"
+    if "keep_default_na" not in kwargs:
+        kwargs["keep_default_na"] = False
+    if "na_values" not in kwargs:
+        kwargs["na_values"] = NA_VALUES
+
+    if os.stat(file).st_size > 0:
+        return pd.read_csv(file, **kwargs)
+    else:
+        return pd.DataFrame()
+
+
+def save_to_geojson(df, fn):
+    """
+    Save a (Geo)DataFrame to a GeoJSON file.
+
+    Parameters:
+    - df: The (Geo)DataFrame to be saved.
+    - fn: The filename (including the path) of the output GeoJSON file.
+
+    Returns:
+    None
+    """
+    if os.path.exists(fn):
+        os.unlink(fn)  # remove file if it exists
+
+    # save file if the (Geo)DataFrame is non-empty
+    if df.empty:
+        # create empty file to avoid issues with snakemake
+        with open(fn, "w") as fp:
+            pass
+    else:
+        # save file
+        df.to_file(fn, driver="GeoJSON")
+
+
+def read_geojson(fn, cols=[], dtype=None, crs="EPSG:4326"):
+    """
+    Function to read a geojson file fn. When the file is empty, then an empty
+    GeoDataFrame is returned having columns cols, the specified crs and the
+    columns specified by the dtype dictionary it not none.
+
+    Parameters:
+    ------------
+    fn : str
+        Path to the file to read
+    cols : list
+        List of columns of the GeoDataFrame
+    dtype : dict
+        Dictionary of the type of the object by column
+    crs : str
+        CRS of the GeoDataFrame
+    """
+    # if the file is non-zero, read the geodataframe and return it
+    if os.path.getsize(fn) > 0:
+        return gpd.read_file(fn)
+    else:
+        # else return an empty GeoDataFrame
+        df = gpd.GeoDataFrame(columns=cols, geometry=[], crs=crs)
+        if isinstance(dtype, dict):
+            for k, v in dtype.items():
+                df[k] = df[k].astype(v)
+        return df
```

Comment:
Similar here. Let's explore if this is really necessary or if it can be replaced by `gpd.read_file(fn)`